pkg/server/server.go

```diff
@@ -0,0 +1,255 @@
+package server
+
+import (
+	"context"
+	"crypto/tls"
+	"crypto/x509"
+	"fmt"
+	"log/slog"
+	"net"
+	"os"
+	"runtime"
+	"sync"
+	"time"
+
+	jobv1 "github.com/kralicky/jobserver/pkg/apis/job/v1"
+	"github.com/kralicky/jobserver/pkg/auth"
+	"github.com/kralicky/jobserver/pkg/jobs"
+	"github.com/kralicky/jobserver/pkg/rbac"
+	"google.golang.org/grpc"
+	"google.golang.org/grpc/codes"
+	"google.golang.org/grpc/credentials"
+	"google.golang.org/grpc/keepalive"
+	"google.golang.org/grpc/status"
+	"google.golang.org/protobuf/types/known/emptypb"
+)
+
+type Options struct {
+	ListenAddress   string
+	CaCertFile      string
+	CertFile        string
+	KeyFile         string
+	AuthMiddlewares []auth.Middleware
+}
+
+type jobInfo struct {
+	jobs.Process
+	owner  auth.AuthenticatedUser
+	cancel context.CancelCauseFunc
+}
+
+type Server struct {
+	Options
+	jobv1.UnsafeJobServer
+	jobs    sync.Map // map[string]jobInfo
+	runtime jobs.Runtime
+}
+
+func NewServer(runtime jobs.Runtime, options Options) *Server {
+	return &Server{
+		Options: options,
+		runtime: runtime,
+	}
+}
+
+type userJobId struct {
+	*jobv1.JobId
+	user auth.AuthenticatedUser
+}
+
+func (i userJobId) AssignedUser() auth.AuthenticatedUser {
+	return i.user
+}
+
+// ListJobs implements v1.JobServer.
+func (s *Server) List(ctx context.Context, _ *emptypb.Empty) (*jobv1.JobIdList, error) {
+	var jobIds []userJobId
+	s.jobs.Range(func(k, v any) bool {
+		jobIds = append(jobIds, userJobId{
+			JobId: &jobv1.JobId{Id: k.(string)},
+			user:  v.(jobInfo).owner,
+		})
+		return true
+	})
+
+	var err error
+	jobIds, err = rbac.FilterByScope(ctx, jobIds)
+	if err != nil {
+		return nil, err
+	}
+
+	var items []*jobv1.JobId
+	for _, id := range jobIds {
+		items = append(items, id.JobId)
+	}
+	return &jobv1.JobIdList{
+		Items: items,
+	}, nil
+}
+
+func (s *Server) lookupScoped(ctx context.Context, id *jobv1.JobId) (jobInfo, error) {
+	var user auth.AuthenticatedUser
+	// if the job doesn't exist, don't short circuit
+	job, ok := s.jobs.Load(id.Id)
+	if ok {
+		user = job.(jobInfo).owner
+	}
+	if err := rbac.VerifyScopeForUser(ctx, user); err != nil {
+		return jobInfo{}, err
+	}
+	if !ok {
+		return jobInfo{}, status.Errorf(codes.NotFound, "job %s not found", id.Id)
+	}
+	return job.(jobInfo), nil
+}
+
+// GetStatus implements v1.JobServer.
+func (s *Server) Status(ctx context.Context, id *jobv1.JobId) (*jobv1.JobStatus, error) {
+	job, err := s.lookupScoped(ctx, id)
+	if err != nil {
+		return nil, err
+	}
+	return job.Status(), nil
+}
+
+// Start implements v1.JobServer.
+func (s *Server) Start(ctx context.Context, in *jobv1.JobSpec) (*jobv1.JobId, error) {
+	user := auth.AuthenticatedUserFromContext(ctx)
+	jobCtx, cancel := context.WithCancelCause(context.TODO())
```

Comment:
I figured `context.WithoutCancel(ctx)` might imply there is some information in the context that would be relevant to pass along, which isn't the case. `context.Background()` also makes sense here, although on the other hand `context.TODO()` sticks out; it seems more deliberate to use. I don't mind either way if you'd prefer I change it. :)

Reply:
`context.TODO` is specifically for situations in which you plan to add a context after some further code changes in the future. The job context here seems to be very much intended to run in the background until specifically canceled by a deliberate action rather than some cascading cancellation, so `context.Background` is appropriate IMO.